third_party/libunwindstack/PeCoffUnwindInfoUnwinderX86_64.cpp

```diff
@@ -0,0 +1,240 @@
+/*
+ * Copyright (C) 2022 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "PeCoffUnwindInfoUnwinderX86_64.h"
+
+#include "unwindstack/MachineX86_64.h"
+#include "unwindstack/Regs.h"
+
+namespace unwindstack {
+
+// The order of registers in PE/COFF unwind information is different from the libunwindstack
+// register order, so we have to map them to the right values. See
+// https://docs.microsoft.com/en-us/cpp/build/exception-handling-x64?view=msvc-160#operation-info
+static uint16_t MapToUnwindstackRegister(uint8_t op_info_register) {
+  std::array<uint16_t, 16> kMachineToUnwindstackRegister = {
+      X86_64_REG_RAX, X86_64_REG_RCX, X86_64_REG_RDX, X86_64_REG_RBX,
+      X86_64_REG_RSP, X86_64_REG_RBP, X86_64_REG_RSI, X86_64_REG_RDI,
+      X86_64_REG_R8,  X86_64_REG_R9,  X86_64_REG_R10, X86_64_REG_R11,
+      X86_64_REG_R12, X86_64_REG_R13, X86_64_REG_R14, X86_64_REG_R15};
+
+  if (op_info_register >= kMachineToUnwindstackRegister.size()) {
+    return X86_64_REG_LAST;
+  }
+
+  return kMachineToUnwindstackRegister[op_info_register];
+}
+
+bool PeCoffUnwindInfoUnwinderX86_64::Eval(Memory* process_memory, Regs* regs,
+                                          const UnwindInfo& unwind_info,
+                                          uint64_t current_code_offset, uint64_t* frame_pointer,
+                                          bool* frame_pointer_used) {
+  // Data is parsed from the object file, so we have to assume that it may be inconsistent.
+  if (unwind_info.num_codes != unwind_info.unwind_codes.size()) {
+    last_error_.code = ERROR_INVALID_COFF;
+    return false;
+  }
+
+  RegsImpl<uint64_t>* cur_regs = reinterpret_cast<RegsImpl<uint64_t>*>(regs);
+
+  for (int op_idx = 0; op_idx < unwind_info.num_codes;) {
+    UnwindCode unwind_code = unwind_info.unwind_codes[op_idx];
+    switch (unwind_code.GetUnwindOp()) {
+      case UWOP_PUSH_NONVOL: {
+        if (unwind_code.code_and_op.code_offset > current_code_offset) {
+          op_idx += 1;
+          continue;
+        }
```

Comment:
I think this is to skip through the parts of the unwinding that is not applicable since we stopped execution in the middle of the prolog. @ronaldfw: Please correct if this is wrong....

I'd mention the "Stopped in prolog" case in the function description in the header.

Reply:
Yes, it's when the prolog hasn't fully executed that we skip unwind operations. In the middle of a function (not in the prolog, not in the epilog) one executes all unwind operations (the epilog is handled differently, to be implemented in one of the next PRs). 

> I assume the if-continue-statements provide that because the code_offsets are always ordered decreasingly in the unwind_codes array, right?

Yes, that is correct.

Example from an actual .dll in the order in which they appear (and are executed in this loop):

```
    Unwind Codes:
      0x0a: UOP_AllocSmall 48
      0x06: UOP_PushNonVol R12
      0x04: UOP_PushNonVol R13
      0x02: UOP_PushNonVol R14
```

And the prolog disassembly for illustration: 

```
    6eba5fa0:	41 56                	push   r14
    6eba5fa2:	41 55                	push   r13
    6eba5fa4:	41 54                	push   r12
    6eba5fa6:	48 83 ec 30          	sub    rsp,0x30
```

Reply:
Also added a brief explanation about skipping of codes in the header as suggested.

Reply:
Cool. Thanks. Makes sense!